### PR TITLE
add openssl support

### PR DESCRIPTION
**Pull request by Stebalien**
Also adds conformance tests to make sure we're generating the same thing each
time.

Also removes:

1. RSA Encrypt. We can't support this with openssl, at least not easily. We also don't really support it generally as the security of reusing the same RSA key for encrypting and signing is questionable.
2. `MarshalRsa{Private,Public}Key`. These didn't return errors but openssl may.

fixes https://github.com/libp2p/go-libp2p-secio/issues/42

### CODE REVIEW

File: openssl_common.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+package crypto
+
+import (
+	pb "github.com/libp2p/go-libp2p-crypto/pb"
+
+	openssl "github.com/spacemonkeygo/openssl"
+)
+
+// define these as separate types so we can add more key types later and reuse
+// code.
```
Comment (Stebalien) [inline openssl_common.go]: Unfortunately, RSA is the only place where we use the PKCS1v15 signature format, as far as I can tell and I can't find an easy way to get something else from this openssl library.

**Comment by marten-seemann** [comment]
The OpenSSL library uses cgo, which afaik will cause the whole binary to be built on cgo mode, which prevents a lot of optimizations usually performed by the Go compiler.

Do we really need to fix this issue, considering that we’re in the procreas of phasing our secio right at this moment?

**Comment by Stebalien** [comment]
This is only enabled when `-tags openssl` is passed to the compiler.

> The OpenSSL library uses cgo, which afaik will cause the whole binary to be built on cgo mode, which prevents a lot of optimizations usually performed by the Go compiler.

I know performance from go to C will suck but I've never heard of it preventing other optimizations.

File: .travis.yml [hunk 2]
```diff
@@ -9,9 +9,10 @@ go:
 env:
   global:
     - GOTFLAGS="-race"
-  matrix:
     - BUILD_DEPTYPE=gomod
-
+  matrix:
+    - GOTFLAGS="-race -tags=openssl"
+    - GOTFLAGS="-tags=openssl"
```
Comment (vyzo) [inline .travis.yml]: shouldn't we have a build in the matrix to test the go implementation as well?

**Comment by marten-seemann** [comment]
Are there any other hosts than mars that are experiencing that problem and that are expected to compile their own IPFS? If not, this might be better left in a separate branch, instead of adding it to the master branch just to become technical debt as soon as the transition to TLS is completed.

> I've never heard of it preventing other optimizations.

I think I’ve read an article about that some time ago, but I could remember it incorrectly.

**Comment by vyzo** [comment]
@marten-seemann the relays are pretty hammered with connections and spend a lot of time in RSA handshakes.

**Comment by marten-seemann** [comment]
@vyzo My main point is that the benefits of this workaround will be pretty short-lives, as we’re already deploying TLS: The last IPFS release added TLS support (but defaults to secio), and the plan is to switch to TLS by default (with secio as a fallback) with the next release.

**Comment by vyzo** [comment]
We'll have secio with us for a while, not everyone in the network updates promptly.
Besides, the debt is pretty minimal.

**Comment by raulk** [comment]
@marten-seemann yeah, we'll need to support SecIO with RSA keys for a while...

**Comment by raulk** [comment]
Sidenote: so happy we don't have to gx new imports any longer.

**Comment by vyzo** [comment]
I am dying to get this bubbled up and deployed to the relays!

**Comment by raulk** [comment]
@vyzo while this is definitely worth a release, in general we're abusing releases by making one per PR merged (almost). Just do this for now:

```
~/Workbench/protocol/workspace-go-libp2p/go-libp2p-daemon master
❯ go get github.com/libp2p/go-libp2p-crypto@feat/openssl
go: finding github.com/libp2p/go-libp2p v0.0.20
go: finding github.com/libp2p/go-libp2p-kad-dht v0.0.10
go: finding github.com/libp2p/go-libp2p-peerstore v0.0.5
go: finding github.com/libp2p/go-libp2p-autonat-svc v0.0.5
go: finding github.com/libp2p/go-libp2p-kbucket v0.1.1
go: finding github.com/libp2p/go-conn-security-multistream v0.0.2
go: finding github.com/libp2p/go-libp2p-transport-upgrader v0.0.2
go: finding github.com/libp2p/go-libp2p-swarm v0.0.3
go: finding github.com/libp2p/go-libp2p-secio v0.0.3
go: finding github.com/libp2p/go-msgio v0.0.2
go: finding github.com/libp2p/go-libp2p-crypto feat/openssl
go: downloading github.com/libp2p/go-libp2p-crypto v0.0.2-0.20190503032304-848064968c37
go: extracting github.com/libp2p/go-libp2p-crypto v0.0.2-0.20190503032304-848064968c37
```

**Comment by vyzo** [comment]
What's wrong with one release per PR when merging in major new functionality? These are patch releases, not minor increments.

**Comment by vyzo** [comment]
In the meantime I'll deploy relays from a branch, using this branch, to get some early testing.

**Comment by raulk** [comment]
> What's wrong with one release per PR when merging in major new functionality? These are patch releases, not minor increments.

Non-comprehensive list of reasons: too much noise, unpredictable for downstream users, haphazard, difficult to keep tabs on, high risk of making bad/incomplete releases, lack of planning...

Basically: we are abusing patch releases as a way to compensate for lack of continuous integration and nightly builds.

You should have an easy way to run the latest daemon commit pulling in all master -- that's where https://github.com/libp2p/workspace-go-libp2p is headed, but incomplete.

**Comment by vyzo** [comment]
I am not interested in running the daemon locally, but rather deploying it to the relay fleet.
Regardless, I am doing that from a branch.

> Basically: we are abusing patch releases as a way to compensate for lack of continuous integration and nightly builds.

That's heavy. I don't think we are abusing anything by making a patch release for each new major PR merged in. And it makes fleet maintenance much easier.

**Comment by raulk** [comment]
@vyzo you are only thinking about your needs, we should think about users too ;-)

Offloading the conversation to https://github.com/libp2p/go-libp2p/issues/621. The lack of CI creates many deficiencies as pointed out in that issue, and patch releases ain't gonna solve those (it's still a manual process based on developer judgement).

**Comment by vyzo** [comment]
I would argue that our users do care about having the latest patch release be up-to-date with master :)
Anyway, end of discussion here, we have an issue for it now.

File: .travis.yml (hunk 2 above)
Comment (Stebalien) [reply .travis.yml]: (oops)

**Comment by Stebalien** [comment]
> If not, this might be better left in a separate branch, instead of adding it to the master branch just to become technical debt as soon as the transition to TLS is completed.

WRT technical debt, this package isn't changed all that much so I don't expect too much maintenance burden. This change is also mostly invisible so we should be able to just remove it without downstream noticing (we're not extending any interfaces).

**Pull request merged**